source/common/unified_malloc_framework/src/utils/utils.h

```diff
@@ -0,0 +1,111 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#if defined(_WIN32)
+#include <windows.h>
+#else
+#include <stdatomic.h>
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct os_mutex_t;
+
+struct os_mutex_t *util_mutex_create(void);
+void util_mutex_destroy(struct os_mutex_t *mutex);
+int util_mutex_lock(struct os_mutex_t *mutex);
+int util_mutex_unlock(struct os_mutex_t *mutex);
+
+#if defined(_WIN32)
+static __inline unsigned char util_lssb_index(long long value) {
+    unsigned long ret;
+    _BitScanForward64(&ret, value);
+    return (unsigned char)ret;
+}
+static __inline unsigned char util_mssb_index(long long value) {
+    unsigned long ret;
+    _BitScanReverse64(&ret, value);
+    return (unsigned char)ret;
+}
+
+// There is no good way to do atomic_load on windows...
+#define util_atomic_load_acquire(object, dest)                                 \
```

Comment:
C++ has atomic loads implemented directly, so we could change these to be declarations (should be fine for performance if we enable lto) and implement it using the C++ apis in the .cpp file.

Reply:
In C++, you need to declare the variable as std::atomic (similarly as with _Atomic in C, and by the way, _Atomic is not supported by MSVC) to use the atomic operations. Only in C++ 20 you have https://en.cppreference.com/w/cpp/atomic/atomic_ref which can be used on any variable.